### PR TITLE
Separate out staking module into balances and payment

### LINKED ISSUE #211: Runtime: Split staking into staking, accounts, transfer and contracts modules
Staking module has become unwieldy.

- Accounts module should handle accounts: Nonce, Free/ReservedBalance, basic transfer mechanics (non-public), account death, indexing & lookups.
- Transfer module should just provide a public transfer call. Enabled in Polkadot.
- Staking module should handle any staking mechanics, slashing, PoS.
- Contracts should be an (optional) add-on in order to provide smart contract, Code and Storage. Disabled in Polkadot.

Lookups in higher-order modules (democracy, council) should be done through a published trait mechanism.

### PR DESCRIPTION
**Pull request by gavofyork**
Does what it says on the tin. Closes #211.

Despite the apparent extra 300 lines, this actually removes quite a lot of code. With a cunning move of the `PublicAux` type from `consensus` into `system`, we have massively reduced the dependency tree of modules.

Now, we have `system` being the root dependency module: `timestamp`, `consensus` and `balances` are dependent on it. The entire on-chain governance and smart-contract system is dependent only on `balances`, while session-management is dependent on `consensus` and `timestamp`, and `staking` is dependent on `session` and `balances`. It's now possible to use `balances` within a PoA system without needing to bring in PoS concepts.

One upshot of this is the hugely simplified mockups for tests in `democracy`, `contract` and `council` modules. Since they are no longer dependent on `consensus`, `session`, `timestamp` and `staking`, all of those modules can be removed from the mock and everything becomes massively simpler.

99% of this PR is just moved code and renaming to correct references. The only significant code change is balances now including a hook `IsAccountLiquid` which allows staking to impose its Bonding semantics on the `transfer` function.

Needless to say, many storage items will have changed from `Staking XXX` to `Balances XXX`. UI maintainers take note. (cc @jacogr )

### CODE REVIEW

File: substrate/runtime/balances/src/genesis_config.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+// Substrate Demo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Build a staking genesis block.
```
Comment (pepyakin) [inline substrate/runtime/balances/src/genesis_config.rs]: Could you update the comment?

File: substrate/runtime/balances/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+		// `system::AccountNonce` is also deleted if `ReservedBalance` is also zero (it also gets
+		// collapsed to zero if it ever becomes less than `ExistentialDeposit`.
+		pub FreeBalance get(free_balance): default map [ T::AccountId => T::Balance ];
+
+		// The amount of the balance of a given account that is exterally reserved; this can still get
+		// slashed, but gets slashed last of all.
+		//
+		// This balance is a "reserve" balance that other subsystems use in order to set aside tokens
+		// that are still "owned" by the account holder, but which are unspendable. This is different
+		// and wholly unrelated to the `Bondage` system used for staking.
```
Comment (pepyakin) [inline substrate/runtime/balances/src/lib.rs]: Do we need this remark about staking and `Bondage`?

File: substrate/runtime/balances/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+	pub fn lookup_address(a: address::Address<T::AccountId, T::AccountIndex>) -> Option<T::AccountId> {
+		match a {
+			address::Address::Id(i) => Some(i),
+			address::Address::Index(i) => Self::lookup_index(i),
+		}
+	}
+
+	// PUBLIC DISPATCH
+
+	/// Transfer some unlocked staking balance to another staker.
```
Comment (pepyakin) [inline substrate/runtime/balances/src/lib.rs]: I think it's better to avoid terms stak{ing,er}

File: substrate/runtime/balances/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+		pub NextEnumSet get(next_enum_set): required T::AccountIndex;
+		// The enumeration sets.
+		pub EnumSet get(enum_set): default map [ T::AccountIndex => Vec<T::AccountId> ];
+
+		// The "free" balance of a given account.
+		//
+		// This is the only balance that matters in terms of most operations on tokens. It is
+		// alone used to determine the balance when in the contract execution environment. When this
+		// balance falls below the value of `ExistentialDeposit`, then the "current account" is
+		// deleted: specifically, `Bondage` and `FreeBalance`. Furthermore, `OnFreeBalanceZero` callback
```
Comment (pepyakin) [inline substrate/runtime/balances/src/lib.rs]: Should we reword this comment? There is no such thing as `Bondage` now

File: substrate/runtime/council/Cargo.toml [hunk 5]
```diff
@@ -17,14 +17,10 @@ substrate-runtime-io = { path = "../../runtime-io", default_features = false }
 substrate-runtime-support = { path = "../../runtime-support", default_features = false }
 substrate-runtime-primitives = { path = "../primitives", default_features = false }
 substrate-runtime-consensus = { path = "../consensus", default_features = false }
+substrate-runtime-balances = { path = "../balances", default_features = false }
```
Comment (pepyakin) [inline substrate/runtime/council/Cargo.toml]: Should balances be added to the features.std array?

File: substrate/runtime/executive/Cargo.toml [hunk 6]
```diff
@@ -17,6 +17,7 @@ substrate-runtime-system = { path = "../system", default_features = false }
 
 [dev-dependencies]
 substrate-primitives = { path = "../../primitives" }
+substrate-runtime-balances = { path = "../balances" }
```
Comment (pepyakin) [inline substrate/runtime/executive/Cargo.toml]: Should balances be added to the features.std array?

Comment (gavofyork) [reply substrate/runtime/executive/Cargo.toml]: no - it's just a dev dependency so it always has `std`

File: substrate/runtime/balances/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Balances: Handles balances.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "std")]
+extern crate serde;
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply substrate/runtime/balances/src/lib.rs]: Sadly not - deriving `Deserialize` requires it to be around.

[the remark this answers is not in the document]
Comment (pepyakin) [reply substrate/runtime/balances/src/lib.rs]: Hm, weird, before posting this, I `--check`ed everything.

**Comment by pepyakin** [comment]
Looks like https://github.com/paritytech/substrate/pull/629#discussion_r214031355 still not addressed

**Comment by gavofyork** [comment]
image

**Comment by gavofyork** [comment]
@pepyakin ^^

**Pull request merged**